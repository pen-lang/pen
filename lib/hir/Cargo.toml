[package]
name = "hir"
version = "0.1.0"
authors = ["Yota Toyama <raviqqe@gmail.com>"]
edition = "2021"

[dependencies]
fnv = "1"
position = { path = "../position" }
serde = { version = "1", features = ["derive", "rc"] }

[dev-dependencies]
once_cell = "1"
pretty_assertions = "1"
