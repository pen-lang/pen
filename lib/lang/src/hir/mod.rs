pub mod analysis;
mod argument;
mod arithmetic_operation;
mod boolean;
mod boolean_operation;
mod call;
mod declaration;
mod definition;
mod else_branch;
mod equality_operation;
mod expression;
mod if_;
mod if_list;
mod if_type;
mod if_type_branch;
mod lambda;
mod let_;
mod list;
mod list_element;
mod module;
mod none;
mod not_operation;
mod number;
mod operation;
mod order_operation;
mod record_construction;
mod record_deconstruction;
mod record_element;
mod record_update;
mod string;
mod try_operation;
mod type_alias;
mod type_coercion;
mod type_definition;
mod variable;

pub use argument::*;
pub use arithmetic_operation::*;
pub use boolean::*;
pub use boolean_operation::*;
pub use call::*;
pub use declaration::*;
pub use definition::*;
pub use else_branch::*;
pub use equality_operation::*;
pub use expression::*;
pub use if_::*;
pub use if_list::*;
pub use if_type::*;
pub use if_type_branch::*;
pub use lambda::*;
pub use let_::*;
pub use list::*;
pub use list_element::*;
pub use module::*;
pub use none::*;
pub use not_operation::*;
pub use number::*;
pub use operation::*;
pub use order_operation::*;
pub use record_construction::*;
pub use record_deconstruction::*;
pub use record_element::*;
pub use record_update::*;
pub use string::*;
pub use try_operation::*;
pub use type_alias::*;
pub use type_coercion::*;
pub use type_definition::*;
pub use variable::*;
