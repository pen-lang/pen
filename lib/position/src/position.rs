use crate::format;
use serde::{Deserialize, Serialize};
use std::{
    cmp::Ordering,
    fmt::{self, Display, Formatter},
    hash::{Hash, Hasher},
    rc::Rc,
};

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct Position(Rc<PositionInner>);

#[derive(Clone, Debug, Deserialize, Serialize)]
struct PositionInner {
    path: String,
    line_number: usize,
    column_number: usize,
    line: String,
}

impl Position {
    pub fn new(
        path: impl Into<String>,
        line_number: usize,
        column_number: usize,
        line: impl Into<String>,
    ) -> Self {
        Self(
            PositionInner {
                path: path.into(),
                line_number,
                column_number,
                line: line.into(),
            }
            .into(),
        )
    }

    pub fn path(&self) -> &str {
        &self.0.path
    }

    pub fn line_number(&self) -> usize {
        self.0.line_number
    }

    pub fn column_number(&self) -> usize {
        self.0.column_number
    }

    pub fn line(&self) -> &str {
        &self.0.line
    }
}

impl Display for Position {
    fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
        write!(formatter, "{}", format(self))
    }
}

impl Eq for Position {}

impl PartialEq for Position {
    fn eq(&self, _: &Self) -> bool {
        true
    }
}

impl Hash for Position {
    fn hash<H: Hasher>(&self, _: &mut H) {}
}

impl Ord for Position {
    fn cmp(&self, _: &Self) -> Ordering {
        Ordering::Equal
    }
}

impl PartialOrd for Position {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

#[cfg(test)]
mod tests {
    use super::Position;

    #[test]
    fn display() {
        assert_eq!(
            format!("{}", Position::new("file", 1, 1, "x")),
            "file\n1:1:\tx\n    \t^"
        );

        assert_eq!(
            format!("{}", Position::new("file", 1, 2, " x")),
            "file\n1:2:\t x\n    \t ^"
        );
    }
}
