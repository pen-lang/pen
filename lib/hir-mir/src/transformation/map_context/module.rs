use super::context_function_name;
use crate::{
    context::CompileContext,
    transformation::{collection_type, equal_operation, hash_calculation},
    CompileError,
};
use fnv::FnvHashSet;
use hir::{
    analysis::{
        expression_visitor, type_canonicalizer, type_comparability_checker, type_visitor,
        AnalysisError,
    },
    ir::*,
    types,
    types::Type,
};
use position::Position;

pub fn transform(context: &CompileContext, module: &Module) -> Result<Module, CompileError> {
    Ok(Module::new(
        module.type_definitions().to_vec(),
        module.type_aliases().to_vec(),
        module.foreign_declarations().to_vec(),
        module.function_declarations().to_vec(),
        module
            .function_definitions()
            .iter()
            .cloned()
            .chain(
                collect_map_types(context, module)?
                    .into_iter()
                    .map(|type_| transform_map_context_function_definition(context, &type_))
                    .collect::<Result<Vec<_>, _>>()?,
            )
            .collect(),
        module.position().clone(),
    ))
}

fn transform_map_context_function_definition(
    context: &CompileContext,
    map_type: &types::Map,
) -> Result<FunctionDefinition, CompileError> {
    let position = map_type.position();
    let context_type = collection_type::transform_map_context(context, position)?;
    let name = context_function_name(map_type, context.types())?;

    Ok(FunctionDefinition::new(
        &name,
        &name,
        Lambda::new(
            vec![],
            context_type.clone(),
            Call::new(
                Some(compile_context_function_type(&context_type, position).into()),
                Variable::new(
                    &context.configuration()?.map_type.context_function_name,
                    position.clone(),
                ),
                [
                    equal_operation::function::transform(context, map_type.key())?,
                    hash_calculation::function::transform(context, map_type.key())?,
                ]
                .into_iter()
                .chain(
                    if type_comparability_checker::check(
                        map_type.value(),
                        context.types(),
                        context.records(),
                    )? {
                        [
                            equal_operation::function::transform(context, map_type.value())?,
                            hash_calculation::function::transform(context, map_type.value())?,
                        ]
                    } else {
                        [
                            compile_fake_equal_function(position).into(),
                            compile_fake_hash_function(position).into(),
                        ]
                    },
                )
                .collect(),
                position.clone(),
            ),
            position.clone(),
        ),
        None,
        false,
        position.clone(),
    ))
}

fn compile_context_function_type(context_type: &Type, position: &Position) -> types::Function {
    let any_type = Type::from(types::Any::new(position.clone()));
    let equal_function_type = Type::from(types::Function::new(
        vec![any_type.clone(), any_type.clone()],
        types::Boolean::new(position.clone()),
        position.clone(),
    ));
    let hash_function_type = Type::from(types::Function::new(
        vec![any_type],
        types::Number::new(position.clone()),
        position.clone(),
    ));

    types::Function::new(
        vec![
            equal_function_type.clone(),
            hash_function_type.clone(),
            equal_function_type,
            hash_function_type,
        ],
        context_type.clone(),
        position.clone(),
    )
}

fn compile_fake_equal_function(position: &Position) -> Lambda {
    Lambda::new(
        vec![
            Argument::new("", types::Any::new(position.clone())),
            Argument::new("", types::Any::new(position.clone())),
        ],
        types::Boolean::new(position.clone()),
        Boolean::new(false, position.clone()),
        position.clone(),
    )
}

fn compile_fake_hash_function(position: &Position) -> Lambda {
    Lambda::new(
        vec![Argument::new("", types::Any::new(position.clone()))],
        types::Number::new(position.clone()),
        Number::new(0.0, position.clone()),
        position.clone(),
    )
}

fn collect_map_types(
    context: &CompileContext,
    module: &Module,
) -> Result<FnvHashSet<types::Map>, AnalysisError> {
    let mut map_types = FnvHashSet::default();

    type_visitor::visit(module, |type_| {
        if let Type::Map(map_type) = type_ {
            map_types.insert(map_type.clone());
        }
    });

    expression_visitor::visit(module, |expression| match expression {
        Expression::IfMap(if_) => {
            map_types.insert(types::Map::new(
                if_.key_type().unwrap().clone(),
                if_.value_type().unwrap().clone(),
                if_.position().clone(),
            ));
        }
        Expression::Map(map) => {
            map_types.insert(types::Map::new(
                map.key_type().clone(),
                map.value_type().clone(),
                map.position().clone(),
            ));
        }
        _ => {}
    });

    map_types
        .into_iter()
        .map(|type_| {
            Ok(types::Map::new(
                type_canonicalizer::canonicalize(type_.key(), context.types())?,
                type_canonicalizer::canonicalize(type_.value(), context.types())?,
                type_.position().clone(),
            ))
        })
        .collect::<Result<_, _>>()
}

#[cfg(test)]
mod tests {
    use super::*;
    use hir::test::{FunctionDefinitionFake, ModuleFake};
    use position::test::PositionFake;

    #[test]
    fn transform_none_key_and_none_value() {
        let context = CompileContext::dummy(Default::default(), Default::default());

        insta::assert_debug_snapshot!(transform(
            &context,
            &Module::empty().set_function_definitions(vec![FunctionDefinition::fake(
                "f",
                Lambda::new(
                    vec![],
                    collection_type::transform_map_context(&context, &Position::fake()).unwrap(),
                    Map::new(
                        types::None::new(Position::fake()),
                        types::None::new(Position::fake()),
                        vec![],
                        Position::fake()
                    ),
                    Position::fake()
                ),
                false
            )])
        ));
    }

    #[test]
    fn transform_function_value() {
        let context = CompileContext::dummy(Default::default(), Default::default());

        insta::assert_debug_snapshot!(transform(
            &context,
            &Module::empty().set_function_definitions(vec![FunctionDefinition::fake(
                "f",
                Lambda::new(
                    vec![],
                    collection_type::transform_map_context(&context, &Position::fake()).unwrap(),
                    Map::new(
                        types::None::new(Position::fake()),
                        types::Function::new(
                            vec![],
                            types::None::new(Position::fake()),
                            Position::fake()
                        ),
                        vec![],
                        Position::fake()
                    ),
                    Position::fake()
                ),
                false
            )])
        ));
    }

    #[test]
    fn do_not_create_duplicate_map_contexts() {
        let context = CompileContext::dummy(
            [
                ("foo".into(), types::None::new(Position::fake()).into()),
                ("bar".into(), types::None::new(Position::fake()).into()),
            ]
            .into_iter()
            .collect(),
            Default::default(),
        );
        let module = Module::empty().set_function_definitions(vec![
            FunctionDefinition::fake(
                "f",
                Lambda::new(
                    vec![],
                    collection_type::transform_map_context(&context, &Position::fake()).unwrap(),
                    Map::new(
                        types::None::new(Position::fake()),
                        types::None::new(Position::fake()),
                        vec![],
                        Position::fake(),
                    ),
                    Position::fake(),
                ),
                false,
            ),
            FunctionDefinition::fake(
                "f",
                Lambda::new(
                    vec![],
                    collection_type::transform_map_context(&context, &Position::fake()).unwrap(),
                    Map::new(
                        types::Reference::new("foo", Position::fake()),
                        types::Reference::new("bar", Position::fake()),
                        vec![],
                        Position::fake(),
                    ),
                    Position::fake(),
                ),
                false,
            ),
        ]);

        assert_eq!(
            transform(&context, &module)
                .unwrap()
                .function_definitions()
                .len()
                - module.function_definitions().len(),
            1,
        );
    }
}
