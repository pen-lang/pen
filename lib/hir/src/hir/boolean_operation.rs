use super::expression::Expression;
use position::Position;
use std::sync::Arc;

#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum BooleanOperator {
    And,
    Or,
}

#[derive(Clone, Debug, PartialEq)]
pub struct BooleanOperation {
    operator: BooleanOperator,
    lhs: Arc<Expression>,
    rhs: Arc<Expression>,
    position: Position,
}

impl BooleanOperation {
    pub fn new(
        operator: BooleanOperator,
        lhs: impl Into<Expression>,
        rhs: impl Into<Expression>,
        position: Position,
    ) -> Self {
        Self {
            operator,
            lhs: lhs.into().into(),
            rhs: rhs.into().into(),
            position,
        }
    }

    pub fn operator(&self) -> BooleanOperator {
        self.operator
    }

    pub fn lhs(&self) -> &Expression {
        &self.lhs
    }

    pub fn rhs(&self) -> &Expression {
        &self.rhs
    }

    pub fn position(&self) -> &Position {
        &self.position
    }
}
