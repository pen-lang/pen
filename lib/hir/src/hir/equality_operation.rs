use super::expression::Expression;
use crate::types::Type;
use position::Position;
use std::sync::Arc;

#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum EqualityOperator {
    Equal,
    NotEqual,
}

#[derive(Clone, Debug, PartialEq)]
pub struct EqualityOperation {
    type_: Option<Type>,
    operator: EqualityOperator,
    lhs: Arc<Expression>,
    rhs: Arc<Expression>,
    position: Position,
}

impl EqualityOperation {
    pub fn new(
        type_: Option<Type>,
        operator: EqualityOperator,
        lhs: impl Into<Expression>,
        rhs: impl Into<Expression>,
        position: Position,
    ) -> Self {
        Self {
            type_,
            operator,
            lhs: Arc::new(lhs.into()),
            rhs: Arc::new(rhs.into()),
            position,
        }
    }

    pub fn type_(&self) -> Option<&Type> {
        self.type_.as_ref()
    }

    pub fn operator(&self) -> EqualityOperator {
        self.operator
    }

    pub fn lhs(&self) -> &Expression {
        &self.lhs
    }

    pub fn rhs(&self) -> &Expression {
        &self.rhs
    }

    pub fn position(&self) -> &Position {
        &self.position
    }
}
