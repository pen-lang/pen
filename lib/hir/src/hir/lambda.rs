use super::{Argument, Expression};
use crate::types::Type;
use position::Position;
use std::sync::Arc;

#[derive(Clone, Debug, PartialEq)]
pub struct Lambda {
    arguments: Vec<Argument>,
    result_type: Type,
    body: Arc<Expression>,
    position: Position,
}

impl Lambda {
    pub fn new(
        arguments: Vec<Argument>,
        result_type: impl Into<Type>,
        body: impl Into<Expression>,
        position: Position,
    ) -> Self {
        Self {
            arguments,
            result_type: result_type.into(),
            body: body.into().into(),
            position,
        }
    }

    pub fn arguments(&self) -> &[Argument] {
        &self.arguments
    }

    pub fn result_type(&self) -> &Type {
        &self.result_type
    }

    pub fn body(&self) -> &Expression {
        &self.body
    }

    pub fn position(&self) -> &Position {
        &self.position
    }
}
