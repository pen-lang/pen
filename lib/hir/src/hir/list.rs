use super::list_element::ListElement;
use crate::types::Type;
use position::Position;

#[derive(Clone, Debug, PartialEq)]
pub struct List {
    type_: Type,
    elements: Vec<ListElement>,
    position: Position,
}

impl List {
    pub fn new(type_: impl Into<Type>, elements: Vec<ListElement>, position: Position) -> Self {
        Self {
            type_: type_.into(),
            elements,
            position,
        }
    }

    pub fn type_(&self) -> &Type {
        &self.type_
    }

    pub fn elements(&self) -> &[ListElement] {
        &self.elements
    }

    pub fn position(&self) -> &Position {
        &self.position
    }
}
