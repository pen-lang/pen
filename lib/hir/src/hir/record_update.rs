use super::{expression::Expression, RecordElement};
use crate::types::Type;
use position::Position;
use std::sync::Arc;

#[derive(Clone, Debug, PartialEq)]
pub struct RecordUpdate {
    type_: Type,
    record: Arc<Expression>,
    elements: Vec<RecordElement>,
    position: Position,
}

impl RecordUpdate {
    pub fn new(
        type_: impl Into<Type>,
        record: impl Into<Expression>,
        elements: Vec<RecordElement>,
        position: Position,
    ) -> Self {
        Self {
            type_: type_.into(),
            record: Arc::new(record.into()),
            elements,
            position,
        }
    }

    pub fn type_(&self) -> &Type {
        &self.type_
    }

    pub fn record(&self) -> &Expression {
        &self.record
    }

    pub fn elements(&self) -> &[RecordElement] {
        &self.elements
    }

    pub fn position(&self) -> &Position {
        &self.position
    }
}
