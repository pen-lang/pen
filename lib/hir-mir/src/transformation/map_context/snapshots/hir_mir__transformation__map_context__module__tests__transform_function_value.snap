---
source: lib/hir-mir/src/transformation/map_context/module.rs
expression: "transform(&context,\n    &Module::empty().set_function_definitions(vec![FunctionDefinition ::\n                fake(\"f\", Lambda ::\n                new(vec! [], collection_type ::\n                transform_map_context(& context, & Position ::\n                fake()).unwrap(), Map ::\n                new(types :: None :: new(Position :: fake()), types ::\n                Function ::\n                new(vec! [], types :: None :: new(Position :: fake()),\n                Position :: fake()), vec! [], Position :: fake()), Position ::\n                fake()), false)]))"
---
Ok(
    Module {
        type_definitions: [],
        type_aliases: [],
        foreign_declarations: [],
        function_declarations: [],
        function_definitions: [
            FunctionDefinition {
                name: "f",
                original_name: "",
                lambda: Lambda {
                    arguments: [],
                    result_type: Reference(
                        Reference {
                            name: "mapContext",
                            position: Position {
                                path: "",
                                line_number: 1,
                                column_number: 1,
                                line: "",
                            },
                        },
                    ),
                    body: Map(
                        Map {
                            key_type: None(
                                None {
                                    position: Position {
                                        path: "",
                                        line_number: 1,
                                        column_number: 1,
                                        line: "",
                                    },
                                },
                            ),
                            value_type: Function(
                                Function {
                                    arguments: [],
                                    result: None(
                                        None {
                                            position: Position {
                                                path: "",
                                                line_number: 1,
                                                column_number: 1,
                                                line: "",
                                            },
                                        },
                                    ),
                                    position: Position {
                                        path: "",
                                        line_number: 1,
                                        column_number: 1,
                                        line: "",
                                    },
                                },
                            ),
                            elements: [],
                            position: Position {
                                path: "",
                                line_number: 1,
                                column_number: 1,
                                line: "",
                            },
                        },
                    ),
                    position: Position {
                        path: "",
                        line_number: 1,
                        column_number: 1,
                        line: "",
                    },
                },
                foreign_definition_configuration: None,
                public: false,
                position: Position {
                    path: "",
                    line_number: 1,
                    column_number: 1,
                    line: "",
                },
            },
            FunctionDefinition {
                name: "hir:map:context:8bbb62926defb736",
                original_name: "hir:map:context:8bbb62926defb736",
                lambda: Lambda {
                    arguments: [],
                    result_type: Reference(
                        Reference {
                            name: "mapContext",
                            position: Position {
                                path: "",
                                line_number: 1,
                                column_number: 1,
                                line: "",
                            },
                        },
                    ),
                    body: Call(
                        Call {
                            function_type: Some(
                                Function(
                                    Function {
                                        arguments: [
                                            Function(
                                                Function {
                                                    arguments: [
                                                        Any(
                                                            Any {
                                                                position: Position {
                                                                    path: "",
                                                                    line_number: 1,
                                                                    column_number: 1,
                                                                    line: "",
                                                                },
                                                            },
                                                        ),
                                                        Any(
                                                            Any {
                                                                position: Position {
                                                                    path: "",
                                                                    line_number: 1,
                                                                    column_number: 1,
                                                                    line: "",
                                                                },
                                                            },
                                                        ),
                                                    ],
                                                    result: Boolean(
                                                        Boolean {
                                                            position: Position {
                                                                path: "",
                                                                line_number: 1,
                                                                column_number: 1,
                                                                line: "",
                                                            },
                                                        },
                                                    ),
                                                    position: Position {
                                                        path: "",
                                                        line_number: 1,
                                                        column_number: 1,
                                                        line: "",
                                                    },
                                                },
                                            ),
                                            Function(
                                                Function {
                                                    arguments: [
                                                        Any(
                                                            Any {
                                                                position: Position {
                                                                    path: "",
                                                                    line_number: 1,
                                                                    column_number: 1,
                                                                    line: "",
                                                                },
                                                            },
                                                        ),
                                                    ],
                                                    result: Number(
                                                        Number {
                                                            position: Position {
                                                                path: "",
                                                                line_number: 1,
                                                                column_number: 1,
                                                                line: "",
                                                            },
                                                        },
                                                    ),
                                                    position: Position {
                                                        path: "",
                                                        line_number: 1,
                                                        column_number: 1,
                                                        line: "",
                                                    },
                                                },
                                            ),
                                            Function(
                                                Function {
                                                    arguments: [
                                                        Any(
                                                            Any {
                                                                position: Position {
                                                                    path: "",
                                                                    line_number: 1,
                                                                    column_number: 1,
                                                                    line: "",
                                                                },
                                                            },
                                                        ),
                                                        Any(
                                                            Any {
                                                                position: Position {
                                                                    path: "",
                                                                    line_number: 1,
                                                                    column_number: 1,
                                                                    line: "",
                                                                },
                                                            },
                                                        ),
                                                    ],
                                                    result: Boolean(
                                                        Boolean {
                                                            position: Position {
                                                                path: "",
                                                                line_number: 1,
                                                                column_number: 1,
                                                                line: "",
                                                            },
                                                        },
                                                    ),
                                                    position: Position {
                                                        path: "",
                                                        line_number: 1,
                                                        column_number: 1,
                                                        line: "",
                                                    },
                                                },
                                            ),
                                            Function(
                                                Function {
                                                    arguments: [
                                                        Any(
                                                            Any {
                                                                position: Position {
                                                                    path: "",
                                                                    line_number: 1,
                                                                    column_number: 1,
                                                                    line: "",
                                                                },
                                                            },
                                                        ),
                                                    ],
                                                    result: Number(
                                                        Number {
                                                            position: Position {
                                                                path: "",
                                                                line_number: 1,
                                                                column_number: 1,
                                                                line: "",
                                                            },
                                                        },
                                                    ),
                                                    position: Position {
                                                        path: "",
                                                        line_number: 1,
                                                        column_number: 1,
                                                        line: "",
                                                    },
                                                },
                                            ),
                                        ],
                                        result: Reference(
                                            Reference {
                                                name: "mapContext",
                                                position: Position {
                                                    path: "",
                                                    line_number: 1,
                                                    column_number: 1,
                                                    line: "",
                                                },
                                            },
                                        ),
                                        position: Position {
                                            path: "",
                                            line_number: 1,
                                            column_number: 1,
                                            line: "",
                                        },
                                    },
                                ),
                            ),
                            function: Variable(
                                Variable {
                                    name: "newMapContext",
                                    position: Position {
                                        path: "",
                                        line_number: 1,
                                        column_number: 1,
                                        line: "",
                                    },
                                },
                            ),
                            arguments: [
                                Variable(
                                    Variable {
                                        name: "hir:equal:d409c28886113c9b",
                                        position: Position {
                                            path: "",
                                            line_number: 1,
                                            column_number: 1,
                                            line: "",
                                        },
                                    },
                                ),
                                Variable(
                                    Variable {
                                        name: "hir:hash:d409c28886113c9b",
                                        position: Position {
                                            path: "",
                                            line_number: 1,
                                            column_number: 1,
                                            line: "",
                                        },
                                    },
                                ),
                                Lambda(
                                    Lambda {
                                        arguments: [
                                            Argument {
                                                name: "",
                                                type_: Any(
                                                    Any {
                                                        position: Position {
                                                            path: "",
                                                            line_number: 1,
                                                            column_number: 1,
                                                            line: "",
                                                        },
                                                    },
                                                ),
                                            },
                                            Argument {
                                                name: "",
                                                type_: Any(
                                                    Any {
                                                        position: Position {
                                                            path: "",
                                                            line_number: 1,
                                                            column_number: 1,
                                                            line: "",
                                                        },
                                                    },
                                                ),
                                            },
                                        ],
                                        result_type: Boolean(
                                            Boolean {
                                                position: Position {
                                                    path: "",
                                                    line_number: 1,
                                                    column_number: 1,
                                                    line: "",
                                                },
                                            },
                                        ),
                                        body: Boolean(
                                            Boolean {
                                                value: false,
                                                position: Position {
                                                    path: "",
                                                    line_number: 1,
                                                    column_number: 1,
                                                    line: "",
                                                },
                                            },
                                        ),
                                        position: Position {
                                            path: "",
                                            line_number: 1,
                                            column_number: 1,
                                            line: "",
                                        },
                                    },
                                ),
                                Lambda(
                                    Lambda {
                                        arguments: [
                                            Argument {
                                                name: "",
                                                type_: Any(
                                                    Any {
                                                        position: Position {
                                                            path: "",
                                                            line_number: 1,
                                                            column_number: 1,
                                                            line: "",
                                                        },
                                                    },
                                                ),
                                            },
                                        ],
                                        result_type: Number(
                                            Number {
                                                position: Position {
                                                    path: "",
                                                    line_number: 1,
                                                    column_number: 1,
                                                    line: "",
                                                },
                                            },
                                        ),
                                        body: Number(
                                            Number {
                                                value: 0.0,
                                                position: Position {
                                                    path: "",
                                                    line_number: 1,
                                                    column_number: 1,
                                                    line: "",
                                                },
                                            },
                                        ),
                                        position: Position {
                                            path: "",
                                            line_number: 1,
                                            column_number: 1,
                                            line: "",
                                        },
                                    },
                                ),
                            ],
                            position: Position {
                                path: "",
                                line_number: 1,
                                column_number: 1,
                                line: "",
                            },
                        },
                    ),
                    position: Position {
                        path: "",
                        line_number: 1,
                        column_number: 1,
                        line: "",
                    },
                },
                foreign_definition_configuration: None,
                public: false,
                position: Position {
                    path: "",
                    line_number: 1,
                    column_number: 1,
                    line: "",
                },
            },
        ],
        position: Position {
            path: "",
            line_number: 1,
            column_number: 1,
            line: "",
        },
    },
)
